[#718] Added Signing Request model

sam-bertin opened the pull request:
- Added the model
- Adding a Signing Request will generate the PK/CSR.
- CSR is displayed / PK is just stored
- CSR and PK stored as textfields under the SR model
- Added feature in documentation
- Model method unit tests

sam-bertin committed:
[#718] Added Signing Request model

- Added the model
- Adding a Signing Request will generate the PK/CSR.
- CSR is displayed / PK is just stored
- CSR and PK stored as textfields under the SR model
- Added feature in documentation
- Model method unit tests

sergei-maertens commented on docs/quickstart.rst:
```diff
@@ -34,6 +34,8 @@ configuration.
 Whenever an instance is deleted (through the admin or code), the associated files are
 purged as well.
 
+You can create and download ``Certificate Signing Request``s (CSR) through the admin as well.
```
```suggestion
You can create and download "Certificate Signing Requests" (CSR) through the admin as well.
```
should fix the failing docs build

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+        with ZipFile(zip_file, "w") as zipf:
+            for csr in queryset:
+                csr_content = csr.csr
+                csr_filename = f"{csr.common_name}_csr.pem"
```
1. I'd call `django.utils.text.slugify` on the common_name too, in the event that someone uses a character like `/` that messes with filenames
2. I suppose it's not impossible that multiple CSRs have the same common name, so to guarantee unique names, you can include the PK

```suggestion
                csr_filename = f"{slugify(csr.common_name)}_{csr.pk}_csr.pem"
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+        zip_file.seek(0)
+        response = HttpResponse(zip_file, content_type="application/zip")
+        response["Content-Disposition"] = 'attachment; filename="csr.zip"'
+        return response
```
I think you can use `django.http.FileResponse` to remove boilerplate, it works especially well for content that can be streamed:

```py
return FileResponse(zip_file, as_attachment=True, filename="csr.zip")
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+        return response
+    elif num == 1:
+        csr = (first := queryset[0]).csr
+        response = HttpResponse(csr, content_type="application/x-pem-file")
```
could probably also use the `FileResponse` approach if you wrap it in a `StringIO` object.

The `FileResponse` also infers the content type header from the filename, so might remove more boilerplate code:

```py
>>> import mimetypes
>>> mimetypes.guess_type("foo.pem")
('application/pem-certificate-chain', None)
```
seems to be fine on my system :)

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+class SigningRequestAdmin(admin.ModelAdmin):
+    fieldsets = (
+        (
+            "Certificate Signing Request Information",
```
```suggestion
            _("Subject information"),
```
please take care to mark all user-facing strings translatable

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+                    "state_or_province_name",
+                    "email_address",
+                ],
+                "description": (
+                    "Fill in this information and click 'SAVE' to generate the CSR."
+                ),
```
```suggestion
                "description": _(
                    "Fill in this information and click 'Save' to generate the CSR."
                ),
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+            },
+        ),
+        (
+            "Certificate Signing Request Content",
```
```suggestion
            _("Signing request"),
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,89 @@
+        (
+            "Certificate Signing Request Information",
+            {
+                "fields": [
```
stylistic/consistency thing - can you use tuples instead of lists for field names?

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -3,11 +3,16 @@
 from django.db import models
 from django.utils.translation import gettext, gettext_lazy as _
 
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
```
can you move the actual `cryptography` usage into a separate module `simple_certmanager/csr_generation.py` and import the functions from there? I'd like to separate the concerns in `models.py` and apply loose coupling.

sergei-maertens replied:
please address this feedback :)

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +21,93 @@
+            # Only store the private key bytes in the text field
+            self.private_key = private_key_file_bytes.decode()
+
+    def __str__(self):
+        return f"Signing Request {self.pk} for {self.common_name}"
```
please mark this string as translatable

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -3,11 +3,16 @@
     Certificate as CryptographyCertificate,
     load_pem_x509_certificate,
 )
+from cryptography.x509.oid import NameOID
```
```suggestion
from cryptography.x509 import NameOID
```
this is the public export - the `.oid` module is not picked up by (Pyright) type checker

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,128 @@
+
+    assert response.status_code == 200
+
+    signing_request = SigningRequest.objects.last()
```
```suggestion
    signing_request = SigningRequest.objects.get()
```
slightly stronger assertion since it will guarantee only one object is created

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,128 @@
+    )
+
+
+def test_generate_private_key():
+    signing_request = SigningRequest(common_name="test.com")
+    assert signing_request.private_key == ""
+    signing_request.generate_private_key()
+    assert signing_request.private_key != ""
+    assert "BEGIN PRIVATE KEY" in signing_request.private_key
```
I'd add an assertion that if additional saves do not overwrite the private key if it's already generated

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,128 @@
+def test_download_csr_single():
+    site = AdminSite()
+    admin = SigningRequestAdmin(SigningRequest, site)
+    request = None  # Simulate a Django request object
```
you can use the `rf` fixture and create an actual request:

```py
request = rf.post("/irrelevant")
```
for the record, I can't tell from this test by itself if it should be POST or GET - if you're testing the admin action, I'd also go through the `admin_client` instead - you can inspect element a bit to get the right data in the post body

if it's GET for a download link in the table, then it's just a simple `admin_client.get(...)` :)

sergei-maertens reviewed (changes requested):
The devil is in the details

The rough lines are fine, just some tweaking to better fit the code (organization) in other libraries of ours

sam-bertin committed:
[#718]Processed PR Feedback

- Make use of FileResponse instead of HttpResponse for file downloads
- Make user content transalatable
- Precisions for filenames, tests, docs

sam-bertin committed:
Fieldset fields ( List => Tuples)

For consistency purpose

sam-bertin committed:
Processing review

- Optimize dowload_csr action / The action's queryset cannot be empty so removing this case helps for better coverage
- added the test for additional save of the csr / does not overwrite the current PK

sam-bertin commented:
@sergei-maertens Yeah thanks for the review. I've adressed your comments. 
I will now try to :
=> make use of FileResponse when dealing with any fille content
=> make use of the _ 'operator' to translate user facing strings
=> use the admin client when testing

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,88 @@
+            },
+        ),
+        (
+            _("Certificate Signing Request Content"),
```
```suggestion
            _("Signing request (CSR)"),
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,88 @@
+                    "email_address",
+                ),
+                "description": (
+                    _("Fill in this information and click 'SAVE' to generate the CSR.")
```
```suggestion
                    _("The CSR will be generated after entering the information and submitting the data.")
```
That uppercase SAVE is a bit shouty, and I think the important part of this description is the answer to "how do I get my CSR???" the question anyway, so this should be an improvement.

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,13 +1,88 @@
+    readonly_fields = ("csr",)
+    actions = [download_csr]
+
+    def response_add(self, request, obj, post_url_continue=None):
+        return HttpResponseRedirect(
+            reverse("admin:simple_certmanager_signingrequest_change", args=(obj.pk,))
+        )
```
you're now replacing the entire default django handling, which breaks the button to "save and add another".

Instead, you'd rather override `response_post_save_add`, which is called by `response_add`.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +21,96 @@
+            self.private_key = private_key_file_bytes.decode()
+
+    def __str__(self):
+        return _("Signing Request #%(pk)s for %(common_name)s") % {
+            "pk": self.pk,
+            "common_name": self.common_name,
+        }
```
can you use the `_("...").format(...)` string interpolation instead? The `%` formatting is legacy.

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,139 @@
+
+@pytest.mark.django_db
+def test_generate_csr():
+    signing_request = SigningRequest(
+        common_name="test.com",
+        country_name="US",
+        organization_name="Test Org",
+        state_or_province_name="Test State",
+        email_address="test@test.com",
+    )
+
+    signing_request.generate_private_key()
+    signing_request.save()
```
```suggestion
    signing_request = SigningRequest.objects.create(
        common_name="test.com",
        country_name="US",
        organization_name="Test Org",
        state_or_province_name="Test State",
        email_address="test@test.com",
    )
```
should be equivalent, since your `def save` takes care of generating the privatey key and CSR.

It was confusing me why the explicit call to generate a private key and additional save call were needed.

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,139 @@
+        },
+    )
+
+    assert isinstance(response, FileResponse)
+    assert response["Content-Type"] == "application/zip"
+    assert "attachment; filename=" in response["Content-Disposition"]
```
cherry on the cake would be taking a CSR out of the zip file and asserting you can load it with `cryptography` to ensure that it didn't get mangled during the process to create a ZIP file :-)

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,139 @@
+        url, {"action": "download_csr", "_selected_action": [signing_request.pk]}
+    )
+
+    assert isinstance(response, FileResponse)
+    assert response["Content-Type"] == "application/pem-certificate-chain"
+    assert "attachment; filename=" in response["Content-Disposition"]
```
I'd also assert that you can load the resulting file with `cryptography`, this covers that we didn't make any mistakes with BytesIO vs. StringIO etc.

sergei-maertens replied:
I think you forgot this one :)

sam-bertin committed:
Process review feedback pt.3

- Updated the help text for the user
- Separated the generation methods from the model
- Added a test CSR to check that It can be read after dowloading as zip
- Updated model and tests to make use of the new module and stop using the model

sergei-maertens commented on simple_certmanager/csr_generation.py:
```diff
@@ -0,0 +1,70 @@
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.x509 import NameOID
+
+# These methods make use of the SigningRequest model to generate a private key and CSR
+# It is not imported here to avoid circular imports
```
This is exactly the part I want to avoid by putting the code in a separate module. This should be pure python and be decoupled entire from Django, since nothing about privatey key/CSR generation is tied to the Django framework.

So, instead of operating on a model instance, you should be taking in the raw primitives, and in the model layer make the changes to call them. The code suggestions below demonstrate a cleaner way to manage this separation of concerns.

sergei-maertens commented on simple_certmanager/csr_generation.py:
```diff
@@ -0,0 +1,70 @@
+
+def generate_csr(signing_request, csr_builder):
+    csr_builder = csr_builder.subject_name(
+        x509.Name(
+            [
+                x509.NameAttribute(NameOID.COUNTRY_NAME, signing_request.country_name),
+                x509.NameAttribute(
+                    NameOID.STATE_OR_PROVINCE_NAME,
+                    signing_request.state_or_province_name,
+                ),
+                x509.NameAttribute(
+                    NameOID.LOCALITY_NAME, signing_request.organization_name
+                ),
+                x509.NameAttribute(
+                    NameOID.ORGANIZATION_NAME, signing_request.organization_name
+                ),
+                x509.NameAttribute(NameOID.COMMON_NAME, signing_request.common_name),
+                x509.NameAttribute(
+                    NameOID.EMAIL_ADDRESS, signing_request.email_address
+                ),
+            ]
+        )
+    )
+
+    return csr_builder
+
+
+def generate_private_key(signing_request):
+    if not signing_request.private_key:
+        # Generate private key
+        private_key = rsa.generate_private_key(
+            public_exponent=65537, key_size=4096, backend=default_backend()
+        )
+        private_key_file_bytes = private_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+        # Only store the private key bytes in the text field
+        signing_request.private_key = private_key_file_bytes.decode()
```
```suggestion
from typing import NewType

from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.x509 import NameOID

PrivateKey = NewType("PrivateKey", str)
CSR = NewType("CSR", str)

# 4096 as "sane default". 2048 is considered too small these days, and the Ansible
# plays generate keys with this key size too.
KEY_SIZE_BITS = 4096


def generate_private_key() -> PrivateKey:
    """
    Generate an RSA private key and return the PEM-encoded key data.
    """
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=KEY_SIZE_BITS,
        backend=default_backend(),
    )
    private_key_file_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    # because it is PEM format, which is base64 encoding, we can safely decode the
    # bytes as ASCII
    return PrivateKey(private_key_file_bytes.decode("ascii"))


def generate_csr(
    key_pem: str,
    *,
    common_name: str,
    country: str = "",
    state_or_province: str = "",
    locality: str = "",
    organization_name: str = "",
    email: str = "",
) -> CSR:
    """
    Generate a CSR for the given private key and return the PEM-encoded CSR data.

    :arg key_pem: The private key to sign the CSR with, provided as PEM-encoded string.
    :arg common_name: The Subject common name, required.
    :arg country: The Subject country, as two-letter country code.
    :arg state_or_province: The Subject state or province name.
    :arg locality: The Subject locality (city) name.
    :arg organization_name: The Subject organization name.
    :arg email: The Subject email address.
    """
    assert common_name, "Common name may not be empty"

    # Load the private key to sign the CSR with.
    private_key = serialization.load_pem_private_key(
        key_pem.encode("ascii"),
        password=None,
        backend=default_backend(),
    )
    # type narrowing - we know we generate RSA private keys, see
    # :func:`generate_private_key`
    assert isinstance(private_key, rsa.RSAPrivateKey)

    # Only include the name attributes in the CSR that have non-empty values
    subject_name_attributes = [
        x509.NameAttribute(attr, value)
        for attr, value in [
            (NameOID.COMMON_NAME, common_name),
            (NameOID.ORGANIZATION_NAME, organization_name),
            (NameOID.LOCALITY_NAME, locality),
            (NameOID.STATE_OR_PROVINCE_NAME, state_or_province),
            (NameOID.COUNTRY_NAME, country),
            (NameOID.EMAIL_ADDRESS, email),
        ]
        if value
    ]
    csr_builder = x509.CertificateSigningRequestBuilder().subject_name(
        x509.Name(subject_name_attributes)
    )

    # Sign the CSR with our private key
    csr = csr_builder.sign(
        private_key, algorithm=hashes.SHA256(), backend=default_backend()
    )

    csr_pem = csr.public_bytes(serialization.Encoding.PEM).decode("ascii")
    return CSR(csr_pem)


def generate_private_key_with_csr(
    *,
    common_name: str,
    country: str = "",
    state_or_province: str = "",
    locality: str = "",
    organization_name: str = "",
    email: str = "",
) -> tuple[PrivateKey, CSR]:
    """
    Generate and return a pair of private key + CSR.

    :arg common_name: The Subject common name, required.
    :arg country: The Subject country, as two-letter country code.
    :arg state_or_province: The Subject state or province name.
    :arg locality: The Subject locality (city) name.
    :arg organization_name: The Subject organization name.
    :arg email: The Subject email address.
    """
    if not common_name:
        raise ValueError("You must provide a non-empty common_name.")

    private_key = generate_private_key()
    csr = generate_csr(
        key_pem=private_key,
        common_name=common_name,
        country=country,
        state_or_province=state_or_province,
        locality=locality,
        organization_name=organization_name,
        email=email,
    )
    return private_key, csr
```

This is roughly equivalent, the changes I make put emphasis on:

* no interaction with django at all
* type annotations
* simple functions with simple primitive arguments
* explicit keyword arguments instead of ambiguous positional arguments
* input validation (common name may not be empty), handling optional values by not including them
* added comments to document some design decisions

Of course, the model layer will need some adaption to interface with these functions, but those are the things that actually belong in the model layer :) See the matching code suggestions there.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -10,12 +10,53 @@
+        blank=True,
+    )
+
+    def save(self, *args, **kwargs):
+        generate_csr_and_private_key(self)
+        super().save(*args, **kwargs)
+
+    def __str__(self):
+        return _("Signing Request #{pk} for {common_name}").format(
+            pk=self.pk,
+            common_name=self.common_name,
+        )
```
we follow django's contributing guidelines to structure models, so the `save` method should be defined after the `__str__`` magic method (see https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#model-style)

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -10,12 +10,53 @@
+        blank=True,
+    )
+
+    def save(self, *args, **kwargs):
+        generate_csr_and_private_key(self)
+        super().save(*args, **kwargs)
```
with my proposed changes in `csr_generation.py`, this becomes:

```suggestion
    def save(self, *args, **kwargs):
        # we generate private key + csr as one atomic unit
        # TODO: add a check-constraint at the DB level to enforce this.
        if not self.private_key:
            self.private_key, self.csr = generate_private_key_with_csr(
                common_name=self.common_name,
                organization_name=self.organization_name,
                # XXX field for locality/city is missing!
                # locality=self.locality_name,
                state_or_province=self.state_or_province_name,
                country=self.country_name,
                email=self.email_address,
            )

        super().save(*args, **kwargs)
```

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -0,0 +1,164 @@
+
+@pytest.mark.django_db
+def test_download_csr_multiple(admin_client):
+    csr_content = open("tests/data/valid_csr.pem").read()
+    signing_request1 = SigningRequest.objects.create(
+        common_name="Test", csr=csr_content
+    )
+    signing_request2 = SigningRequest.objects.create(
+        common_name="Test2", csr=csr_content
+    )
```
mm, what's the reason for including a fixed file here instead of just letting the private key + csr generation do its job?

I'd expected something like this instead:

```py
SigningRequest.obejcts.create(common_name="test.com", country_name="NL")
```
and that should be baked into the CSR.

sam-bertin replied:
I did it this way cause I only wanted to test the end of the process (the download) rather than testing the full process but yeah I think it's good to have a full integration test instead.

sergei-maertens replied:
I wouldn't say we are "testing the full process", but rather we are using the new feature in the test setup. the assertions are all about the zip download and that's good :)

sam-bertin committed:
Processed Review

- Full decoupling of the csr_generation methods.
 => (
    no interaction with django at all
    type annotations
    simple functions with simple primitive arguments
    explicit keyword arguments instead of ambiguous positional arguments
    input validation (common name may not be empty), handling optional values by not including them
    added comments to document some design decisions)
- Code styles
- Added Locality field onto the CSR
- Added a Constraint for CSR and PK to bet set up atomically
- Removing unused mocked file valid_csr

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -10,12 +10,76 @@
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(
+                    models.Q(csr__isnull=False, private_key__isnull=False)
+                    | models.Q(csr__isnull=True, private_key__isnull=True),
+                ),
```
ah this isn't entirely it - because they are text fields, they will never be `null` (unless you set `null=True` on the model fields, but we specifically do not allow this on text fields)

The correct check would be something like:

```py
models.Q(csr="", private_key="") 
| (~models.Q(csr="") & ~models.Q(private_key=""))
```
